File: src/llm-cost.ts
import type { CompletionUsage } from 'openai/resources.js'
import type { SupportedChatModel } from './schema.js'

// Model pricing data
type ModelPricing = {
  inputCostPerMillion: number
  outputCostPerMillion: number
}

const MODEL_PRICING: Partial<Record<SupportedChatModel, ModelPricing>> = {
  'gpt-5.2': {
    inputCostPerMillion: 1.75,
    outputCostPerMillion: 14.0,
  },
  'gemini-2.5-pro': {
    inputCostPerMillion: 1.25,
    outputCostPerMillion: 10.0,
  },
  'gemini-3-pro-preview': {
    inputCostPerMillion: 2.0,
    outputCostPerMillion: 12.0,
  },
  'gemini-3.1-pro-preview': {
    inputCostPerMillion: 2.0,
    outputCostPerMillion: 12.0,
  },
  'deepseek-reasoner': {
    inputCostPerMillion: 0.55,
    outputCostPerMillion: 2.19,
  },
}

export function calculateCost(
  usage: CompletionUsage | undefined,
  model: SupportedChatModel,
): { inputCost: number; outputCost: number; totalCost: number } {
  const pricing = MODEL_PRICING[model]
  if (!pricing) {
    return { inputCost: 0, outputCost: 0, totalCost: 0 }
  }

  const inputTokens = usage?.prompt_tokens || 0
  const outputTokens = usage?.completion_tokens || 0
  const inputCost = (inputTokens / 1_000_000) * pricing.inputCostPerMillion
  const outputCost = (outputTokens / 1_000_000) * pricing.outputCostPerMillion
  const totalCost = inputCost + outputCost

  return { inputCost, outputCost, totalCost }
}

File: src/llm_cost.rs
struct ModelPricing {
    input_per_million: f64,
    output_per_million: f64,
}

pub struct CostResult {
    pub input_cost: f64,
    pub output_cost: f64,
    pub total_cost: f64,
}

pub fn calculate_cost(prompt_tokens: u64, completion_tokens: u64, model: &str) -> CostResult {
    let pricing = match model {
        "gpt-5.4" => Some(ModelPricing {
            input_per_million: 2.5,
            output_per_million: 15.0,
        }),
        "gpt-5.2" => Some(ModelPricing {
            input_per_million: 1.75,
            output_per_million: 14.0,
        }),
        "gemini-2.5-pro" => Some(ModelPricing {
            input_per_million: 1.25,
            output_per_million: 10.0,
        }),
        "gemini-3-pro-preview" => Some(ModelPricing {
            input_per_million: 2.0,
            output_per_million: 12.0,
        }),
        "gemini-3.1-pro-preview" => Some(ModelPricing {
            input_per_million: 2.0,
            output_per_million: 12.0,
        }),
        "deepseek-reasoner" => Some(ModelPricing {
            input_per_million: 0.55,
            output_per_million: 2.19,
        }),
        _ => None,
    };

    match pricing {
        Some(p) => {
            let input_cost = (prompt_tokens as f64 / 1_000_000.0) * p.input_per_million;
            let output_cost = (completion_tokens as f64 / 1_000_000.0) * p.output_per_million;
            CostResult {
                input_cost,
                output_cost,
                total_cost: input_cost + output_cost,
            }
        }
        None => CostResult {
            input_cost: 0.0,
            output_cost: 0.0,
            total_cost: 0.0,
        },
    }
}
